Detokenizer fixes

jaime-m-p opened the pull request:
This PR tries to solve most common problems with detokenization:
 - spaces after special tokens
 - spaces before and after apostrophes

Related issues: https://github.com/ggerganov/llama.cpp/issues/8023, https://github.com/ggerganov/llama.cpp/issues/7938.

___
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

Commit:
Add llama_detokenize()

Commit:
Using llama_tokenize() in tests

Commit:
Using llama_tokenize() in tests

Commit:
Fix tokenizer tests

Commit:
minor: confusing hexadecimal codepoint

Commit:
Clean old known problematic codepoints

Commit:
Update bruteforce random tests

Add detokenizer checks
New generator: ascii_lr_strip
New generator: apostrophe
Add more vocabs files

Commit:
Fix add_space_prefix, set false by default

Commit:
Remove previous space

Commit:
Remove previous space

Commit:
Add tokenizer flag: clean_up_tokenization_spaces

cmp-nct commented:
I gave it a test run using phi3-mini-instruct
First I ran a string through the tokenize_test.py which the model came with:
Both of those two strings tokenize identical (the \n is ignore for some reason)
`Test:<|user|>This is a test<|end|>\n<|assistant|>`  
`Test:<|user|>This is a test<|end|><|assistant|>`
```
ID: 1, Content: '<s>'
ID: 4321, Content: '▁Test'
ID: 29901, Content: ':'
ID: 32010, Content: '<|user|>'
ID: 910, Content: '▁This'
ID: 338, Content: '▁is'
ID: 263, Content: '▁a'
ID: 1243, Content: '▁test'
ID: 32007, Content: '<|end|>'
ID: 32001, Content: '<|assistant|>'
```

Now I ran the same through llama.cpp tokenization:
```
<|assistant|>':
  4321 -> ' Test'
 29901 -> ':'
 32010 -> '<|user|>'
   910 -> ' This'
   338 -> ' is'
   263 -> ' a'
  1243 -> ' test'
 32007 -> '<|end|>'
 29871 -> ' '
    13 -> '
'
 32001 -> '<|assistant|>'
```

It looks like the python tokenizer does not add anything between two special tokens.

Commit:
tests: unexpected vocab type as test fail instead of error

Useful when automating tests:
 - If you don't know in advance the vocab type.
 - Differenciate other loading errors.

Commit:
tests: gracefully exit threads

Using exit() is throwing random exceptions

Commit:
tets: skip unicode surrogaes and undefined

Commit:
Fix detokenizer():

UNKNOWN and CONTROL are 'special pieces'.
Remove space after UNKNOWN and CONTROL.
Refactor llama_token_to_piece().

cmp-nct commented:
@jaime-m-p 
Oh, I didn't know about that behavior. Explains a lot!
1) The \n was manually added by me based on the official template by Microsoft
![image](https://github.com/ggerganov/llama.cpp/assets/78893154/1601d7c5-39ce-4eeb-8e6a-b1fd438554ca)
But as you pointed out: END has "rstrip" true, so the newline is removed.

2) model name special behavior
That's great to know. It also explains why it's not working for me!
I am testing this on a recent xtuner phi3-mini model (llava type) which comes with genetic.name == "LLaMA v2"

I'll repeat the tests after fixing those issues and reverting my changes. Given your results that's promising.

It's a little troublesome that such errors can very easily sneak into a model and it's very hard to notice them, even harder to fix them without blindly recreating the model from originals.
Likely the majority of GGUF files on huggingspace are flawed in some way.

Harsha-Nori commented:
Hi @jaime-m-p and @cmp-nct, really grateful you guys are looking into this! I'm traveling without reliable access to a computer at the moment, but wanted to ask if these fixes now keep stability on retokenization with Phi-3 (i.e. the roundtrip of text -> tokens -> text -> tokens results in the same tokens). The constant whitespace insertion on each cycle was causing serious kv-cache reuse issues on our side and I'm really hopeful that this update resolves it!

Commit:
Do not remove space when decoding special tokens

Commit:
style: remove trailing whitespace

Commit:
Bugfix: custom regexs splits undefined unicode codepoints

Commit:
Update brute force test:

Detokenize special tokens.
Replace errors with '\uFFFD' when detokenizing to 'utf-8'.
More edge cases.
Better detokenization results check.

Commit:
Symetric params for llama_tokenize() and llama_detokenize()

Commit:
Better leading space removal

Commit:
Update bruteforce test

Harsha-Nori commented:
Hmm, great point. I think what I'm really hoping for is eventual stability on the second or third tokenize/detokenize cycles -- before your PR, Phi-3 had the problem of constantly changing the token_id at index 1 (due to growing spaces), which really caused issues.

I think this set of changes is good enough to solve most of our problems :).

Commit:
style : remove spaces

Commit:
Merge commit 'f8c4c073' into detokenizer

Commit:
'viking' detokenizer clean spaces

Commit:
Better leading space removal

Commit:
Update bruteforce test: header files location

Commit:
Update bruteforce test: add more models

Commit:
style: spaces

ggerganov commented on tests/test-tokenizer-1-bpe.cpp:
```diff
@@ -136,6 +137,10 @@ int main(int argc, char **argv) {
             t.join();
         }
+
+        if(errcode) {
```
```suggestion
        if (errcode) {
```

ggerganov reviewed (approved):
Nice work

jaime-m-p committed:
style: spaces

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>
